llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp

```diff
@@ -230,7 +230,11 @@ Type *VPTypeAnalysis::inferScalarType(const VPValue *V) {
             return V->getUnderlyingValue()->getType();
           })
           .Case<VPWidenCastRecipe>(
-              [](const VPWidenCastRecipe *R) { return R->getResultType(); });
+              [](const VPWidenCastRecipe *R) { return R->getResultType(); })
+          .Case<VPExpandSCEVRecipe>([](const VPExpandSCEVRecipe *R) {
```

Comment:
It is needed to infer the type of the step in the new code in  VPlanTransforms.cpp (`if (TypeInfo.inferScalarType(BaseIV) != TypeInfo.inferScalarType(Step)) {)`).

Reply:
VPDerivedIVRecipe getScalarType would be good to remove, will do as follow-up.

`VPWidenIntOrFpInductionRecipe's` trunc drives truncating the start value and step, effectively creating a narrow phi. I'll check to see if this can also be modeled with the new recipe.